### ISSUE TITLE
[privacy] Follow-up to #47: `sklearn.metrics.accuracy_score` is rejected as an API key, and any output ending in a newline is rejected

### ISSUE THREAD

**Issue by YiWang24** [DefuzeX-AI/KUMA-DefuzeX#64]
Follow-up to #47, as invited on closing:

> If rejection persists on current service, please open a new report with request identifiers
> and a minimal sanitized example.

It persists. Two triggers are now isolated to exact strings, both reproducible in a single-step
Run with a body of a few dozen bytes.

## 1. Any token starting with `sk` is treated as an API key

```
output = "sklearn.metrics.accuracy_score"     ->  ValidationError(invalid_request, retryable=False)
output = "sklearn.metrics.cohen_kappa_score"  ->  ValidationError(invalid_request, retryable=False)
output = "xklearn.metrics.cohen_kappa_score"  ->  accepted          <-- one character changed
```

Changing the leading `s` to `x` is the whole difference. The full measured set:

| output body | bytes | result |
|---|---|---|
| `sklearn` | 7 | accepted |
| `sklearn.metrics` | 15 | accepted |
| `sklearn.metrics.accuracy_score` | 32 | **rejected** |
| `sklearn.metrics.cohen_kappa_score` | 33 | **rejected** |
| `xklearn.metrics.cohen_kappa_score` | 33 | accepted |
| `ska1b2c3d4e5f6g7h8i9j` | 21 | **rejected** |
| `ska1b2c3d4e5f6g7h8i9j0` | 22 | **rejected** |
| `statsmodels.stats.inter_rater.fleiss_kappa` | 44 | accepted |
| `I used scikit-learn today` | 25 | accepted |

So the rule is roughly "`sk` followed by enough identifier characters", with the threshold
between 13 and 19 trailing characters. Note what is missing compared to the pattern the SDK
itself ships (`src/kuma/repository/privacy.py`):

```python
("sk_api_key", re.compile(r"\bsk-(?:(?:proj|ant-api03)-)?[A-Za-z0-9_-]{20,}\b")),
```

The SDK requires the **hyphen** after `sk`, and does not admit `.` in the body. The Backend
rule apparently requires neither, so it matches ordinary dotted Python identifiers. Running the
SDK's own scanner over the real payload confirms the disagreement — zero findings against all
eight shipped patterns, with a positive control proving the scanner is live:

```
step-6 findings: 0
positive control findings: 3
```

The practical effect: **an agent that mentions scikit-learn loses its entire multi-step Run.**
`sklearn.metrics.*` is not exotic vocabulary for an agent reasoning about evaluation.

## 2. An output ending in a newline is rejected

Independently of the above, one byte decides it:

```
output = "hello"     ->  accepted
output = "hello\n"   ->  ValidationError(invalid_request, retryable=False)
```

Confirmed on a real payload as well — a 15 KB step that is accepted as produced becomes a
non-retryable failure with a newline appended:

```
step-10 output, 15,317 B, as produced  ->  accepted, verdict returned
step-10 output + "\n", 15,318 B        ->  rejected
```

Stability: a 4-byte body ending in a newline was rejected 3 out of 3 consecutive Runs.

A trailing newline is the most ordinary thing an LLM emits. Any Agent that does not explicitly
`rstrip()` its output loses the Run at the final submission.

## How this was found, and the retest that started it

#47's three Runs could not be reproduced individually because the complete payloads were not
available to you. They were available on our side, recorded step by step. On 2026-09-09 we
replayed one of them byte for byte into fresh Runs — the Agent was not re-executed, so the only
variable between the two experiments is the Backend.

| | 2026-09-08 11:01 UTC | 2026-09-09 09:52 UTC |
|---|---|---|
| arm A | `SensitiveDataError(sensitive_content_detected, retryable=False)` | `ValidationError(invalid_request, retryable=False)` |
| arm B | same | same |
| failing step | step-10 (final) | step-10 (final) |
| `run_id` A | `run_fb9528bc563449ef973163e67b0de7af` | `run_6b84ce69b532487a8fa02af9453fc803` |
| `run_id` B | `run_5f839c8641474f01936f6c01bfe746b3` | `run_f2ae5080361a4123a97e2f958bbc00d3` |

The deployed fix changed the label, not the outcome. `invalid_request` is a Backend code, not a
local guard — it occurs exactly twice in the SDK, both in the Backend error mapping
(`src/kuma/transport/backend.py`):

```python
_VALIDATION_CODES = {..., "invalid_request", ...}
_CODE_ERROR_MESSAGES = {..., "invalid_request": "提交的内容有问题，请检查后重试。"}
```

Bisecting the ten recorded steps of that Run:

```
steps 1..5                        64,974 B  ->  accepted
steps 1..6                        80,987 B  ->  REJECTED at the final step
all ten steps                    139,386 B  ->  REJECTED at the final step

step 6 alone, one-step Case       16,013 B  ->  REJECTED
steps 1-5 + 10, step 6 removed    80,291 B  ->  accepted
step 10 alone                     15,317 B  ->  accepted
```

One step out of ten poisons the whole bundle, because the evidence upload at
`providers/official_judge.py:424 -> _evidence_upload()` carries every step. Narrowing inside
step 6 reached a single 583-byte line, then a single clause:

> `sklearn.metrics.cohen_kappa_score` and `sklearn.metrics.accuracy_score` cover the simpler
> pairwise/agreement cases; `statsmodels.stats.inter_rater.fleiss_kappa` implements Fleiss' κ
> directly on the count matrix `n_{ij}` described in Section 4.

The first clause is rejected; the second is accepted. That is what led to the one-character test
above.

Controls, all on 2026-09-09, same key, same code path, within the hour:

| control | result |
|---|---|
| Backend health probe: read plane, full official chain, verbose-profile casegen, official Judge ×3 | all green, Judge 3/3 |
| Replay of a different Case from the same batch (27,051 B, 9 steps) | verdict, both arms agreeing |
| Replay of the **largest** Case in the same batch (147,033 B, 10 steps) | verdict, both arms agreeing |

Size was our first hypothesis and it is refuted twice over: 147,033 B is accepted, 21 B is
rejected.

## Repro

```python
class Provider:
    agent_profile_required = False

    def generate_case(self, ctx):
        return {"case_id": cid, "input_type": "text",
                "inputs": [{"input_id": "step-1", "payload": "Report status.",
                            "payload_type": "text"}]}

run = create_run(
    repo_path=repo, agent_profile_path=None, case_provider=Provider(),
    judge_provider=None, max_steps=1, judge=True,
    on_failure="continue", track_files=False, save_local=False,
    allow_local=True, api_key=API_KEY, operation_wait_timeout=1800.0,
)
run.get_input(full=True)
run.submit([{"role": "assistant",
             "parts": [{"type": "text", "content": BODY}]}], status="completed")
```

`BODY = "xklearn.metrics.accuracy_score"` returns normally.
`BODY = "sklearn.metrics.accuracy_score"` raises `ValidationError(code=invalid_request,
retryable=False)`. Same for `BODY = "hello\n"` versus `BODY = "hello"`.

One note for reproducing, because it cost us a full invalid run: `judge=True` is required. The
sensitive scan and the evidence upload both hang off
`providers/official_judge.py:424 -> _evidence_upload()`, so a probe built with `judge=False`
skips the failing path entirely and every input passes. Rejected submissions are not billed;
accepted ones are.

## Suggested fix

1. Align the Backend's `sk` rule with the pattern the SDK ships: require the hyphen
   (`sk-`, `sk-proj-`, `sk-ant-api03-`) and exclude `.`. As it stands, any sufficiently long
   `sk`-prefixed identifier — `sklearn` being the obvious one — is a permanent dead end.
2. Accept trailing whitespace, or normalize it before admission, rather than failing a paid,
   completed multi-step Run over a character the Agent author never chose.
3. The request from #47 still stands and is exactly what made these two take a bisection rather
   than a glance: record and return what tripped the check — the matched rule, and the offending
   fragment or its offset. `case_step_limit_exceeded` is the working precedent for a closed,
   fail-closed detail shape that leaks nothing (#57 covers the SDK half).

**Comment by DDai22** [comment]
The server-side fix is now deployed. Agent output preserves its original whitespace, including trailing newlines and nested JSON string whitespace, so validation uses the same content hash as the SDK. Ordinary text such as sklearn.metrics.accuracy_score is no longer rejected solely by an overly broad credential prefix check; protection against actual credentials remains enabled. No SDK hash changes or scanner bypass are needed. Verification used deterministic synthetic cases, not a paid rerun of user requests. Thank you for the reproducible report.
